### DISCUSSION TITLE
UART on pins 16, 17 on FabGL ESP32 Board 16MB development board?

### DISCUSSION THREAD

**Discussion by vitasam** [fdivitto/FabGL#290] [Q&A]
Hello,
I just ordered `FabGL ESP32 Board 16MB Flash 4 MB PSRAM - 3.3V I/O` development board and have a question - is it possible to enable HW UART on pins 16, 17 to use it from PC Emulator as a COM2?

**Comment by fdivitto** [comment]
Hello,
thanks for supporting FabGL project.

On Wrover-E GPIOs 16 and 17 are connected to the module’s integrated PSRAM, so they cannot be used (ref. https://www.espressif.com/sites/default/files/documentation/esp32-wrover-e_esp32-wrover-ie_datasheet_en.pdf , page 11).

If you don't need mouse you may use PS/2 port 1, pins 26 and 27.

[schema.pdf](https://github.com/fdivitto/FabGL/files/9886101/schema.pdf)

**Reply by vitasam** [reply]
Thanks for the information! Regarding PS/2 mouse pins 26 and 27 - they can't be configured as HW UART, so only SW UART implementation is possible?

What will be the best way to get HW UART as COM2 on the development board?

**Comment by fdivitto** [comment]
Unfortunately there aren't available pins (this is the reason for I/O expander). I guess any pin can be configured for hardware UART (I will check this in depth anyway).

**Reply by vitasam** [reply]
Thanks a lot!

**Reply by vitasam** [reply]
> Unfortunately there aren't available pins (this is the reason for I/O expander). I guess any pin can be configured for hardware UART (I will check this in depth anyway).

Hi, do you know can `GPIO0` and `GPIO2` (pins 25, 24) be used as `HW UART`?
![image](https://user-images.githubusercontent.com/5597505/208254514-89aae528-88d8-4698-bc9c-2bfc99627262.png)

According to [datasheet](https://www.espressif.com/sites/default/files/documentation/esp32_datasheet_en.pdf) (page 21), those pins are strapping pins, so the software can read the values of these five bits from register `GPIO_STRAPPING`, but as far as I understand after reset release, the strapping pins work as normal-function pins.

By the way, I was trying to test GPIO0 and GPIO2 pins with [ESP32-SoftwareSerial](https://github.com/junhuanchen/Esp32-SoftwareSerial) library in a simple Arduino script but did not get it working.

**Reply by vitasam** [reply]
It seems that HW UART2 could be used with GPIO0 and GPIO2. I made a simple script:
```
/*
  UART1 and UART2 in a same time
*/

void setup() {
  Serial.begin(9600);
  while (!Serial) {
    ;  // wait for serial port to connect. Needed for native USB port only
  }

  delay(100);

  Serial2.begin(9600, SERIAL_8N1, 0, 2);
  while (!Serial2) {
    ; 
  }

  Serial.println("Hello from UART1");
  delay(100);
  
  Serial2.println("Hello from UART2 (0, 2)");
}

void loop() {
  delay(100);
}
```
`Hello from UART1` was visible, as expected, from USB-COM port. I've connected another UART-to-USB adapter to GPIO0,2 port and it seems to work:
 
![image](https://user-images.githubusercontent.com/5597505/208256902-ff94a620-23c2-4ec0-821b-092c3a6e0ebc.png)

**Reply by vitasam** [reply]
I got UART2 on GPIOs 0,2 working from PC Emulator:
```
  // if (machine->isFabGlDevelopmentBoard())
  {
    auto serial2 = new SerialPort;
    serial2->setSignals(UART2_RX, UART2_TX);
    machine->setCOM2(serial2);
  }
```
 A test program (Turbo Pascal) sends a test string to COM2 through an additional COM2USB adapter:
![turbo_pascal_test_to_COM2](https://user-images.githubusercontent.com/5597505/208268090-da569c5c-ebd3-4fd6-98e4-52fa853bea82.jpg)

<img width="555" alt="uart2_as_COM2" src="https://user-images.githubusercontent.com/5597505/208268120-08874507-2e49-4746-9ec6-3b22b89ae600.PNG">

![com2usb_on_GPIO0_2](https://user-images.githubusercontent.com/5597505/208268140-3016bee4-cb85-4b1e-96ff-18e60084499b.jpg)

@fdivitto what do you think if I make such a pull request - if `machine` class detects FabGL dev-board, the `COM2` port is activated?

P.S. Actually writing from FreeDOS (Pascal) program causes PCEmulator to get strings from both addresses, regardless of the fact it was COM1 or COM2:
![image](https://user-images.githubusercontent.com/5597505/208312353-0d90044c-f080-4a7d-abf6-794cc6596582.png)

**Reply by vitasam** [reply]
P.P.S. Adding SerialPort setup to UART1 and 2 respectively did help - sending a test string from FreeDOS app to COM1 goes to "USB serial port" only. And sendind to COM2 goes to GPIO0,GPIO2 port:

```
  auto serial1 = new SerialPort;
  serial1->setSignals(UART1_RX, UART1_TX);
  serial1->setup(1, 9600, 8, 'N', 1, FlowControl::Software);
  machine->setCOM1(serial1);

  // if (machine->isFabGlDevelopmentBoard())
  //{
    auto serial2 = new SerialPort;
    serial2->setSignals(UART2_RX, UART2_TX);
    serial2->setup(2, 9600, 8, 'N', 1, FlowControl::Software);
    machine->setCOM2(serial2);
  //}
```

**Comment by andromeda2911** [comment]
Pin I/O 26 and 27 (PS/2) from I/O expander is 3.3v TTL or 5v TTL ?

If it's 5v, can I use these pins directly with an arduino 5v ttl?

**Reply by vitasam** [reply]
The board I just ordered, has expander GPIOs that are 3.3v tolerant, according to [datasheet](https://d3s5r33r268y59.cloudfront.net/datasheets/24612/2021-08-11-13-42-32/getting_started_-_devboard_3v3.pdf):
![image](https://user-images.githubusercontent.com/5597505/200410316-8d839ab5-7719-44ec-b23e-f2a58075bb48.png)

**Comment by andromeda2911** [comment]
in your description from version 1.0.9 you say.
AnsiTerminal: use of PS/2 port as serial port (5V levels, RX/TX only).
so pins 26 and 27 in the connector on the right from dev board are indeed 5 V?

**Reply by andromeda2911** [reply]
If I take a PS/2 cable, the end of which I cut, and I connect the PS/2 cable to the fabgl card, at the end of the wires of the PS/2 cable which has been cut, pins 25 and 26 are in 5v?

**Reply by vitasam** [reply]
Are you talking about connector PS/2 pins on `J11` connector ([schematics](https://github.com/fdivitto/FabGL/files/9886101/schema.pdf))? 

If so, they are not 5v tolerant, AFAIK

**Comment by andromeda2911** [comment]
No, connector PS/2 Port1 (Mouse) is J6, bottom right from schematic

**Reply by vitasam** [reply]
J6 ("native PS/2 connector") should be 5V tolerant, I guess, because uC-pins are separated behind resistors

**Reply by andromeda2911** [reply]
yes that's it, but I just realized that it was indicated (5v) on the diagram, so it's interesting to connect a card with 5v TTL serial port to connector PS/2 with cable PS/2.

**Reply by vitasam** [reply]
It depends. If I remember, PS/2 devices do have open-collector interface that has two possible state: low and high impedance. In the "low" state, a transistor pulls the line to ground level.  In the "high impedance" state, the interface acts as an open circuit and doesn't drive the line low or high. So technically there is no 5V on that pins of the board.

**Reply by andromeda2911** [reply]
it's possible, in general it's CLK and DAT, I2C, I thought maybe by making a small circuit it would be possible.

**Reply by andromeda2911** [reply]
this is a PS/2 serial mouse adapter but it may look like this
![61hgP3CFr+L _AC_SL1400_](https://user-images.githubusercontent.com/46875381/208522627-c6754ea7-62a4-44d9-900d-836b4f46b167.jpg)
